File: sage/services/snarc/detectors/novelty.py
"""
Novelty Detector - Memory Comparison

Detects how different current observation is from past experiences.
Maintains episodic memory of sensor observations, computes novelty
as dissimilarity to all past observations.
"""

import time as _time

import torch
import numpy as np
from typing import List, Any, Optional, Tuple
from collections import deque

from sage.services.snarc.temporal import time_decay_weight, DEFAULT_HALF_LIVES


class NoveltyDetector:
    """
    Detect difference from past experiences

    Maintains circular buffer of past observations,
    computes novelty as distance to nearest past experience.
    """

    def __init__(self, memory_size: int = 1000, comparison_samples: int = 50,
                 half_life: float = DEFAULT_HALF_LIVES['novelty']):
        """
        Args:
            memory_size: How many past observations to remember
            comparison_samples: How many recent samples to compare against
                               (for efficiency - don't compare to ALL history)
            half_life: Time-decay half-life in seconds. Old observations feel
                       less familiar (their similarity pull fades with age).
        """
        self.memory_size = memory_size
        self.comparison_samples = comparison_samples
        self.half_life = half_life

        # Episodic memory per sensor: (observation, timestamp) pairs
        self.memory: dict[str, deque] = {}
        self._memory_timestamps: dict[str, deque] = {}

    def compute(self, sensor_output: Any, sensor_id: str) -> float:
        """
        Compute novelty of observation

        Args:
            sensor_output: Current sensor reading
            sensor_id: Unique sensor identifier

        Returns:
            Novelty score (0.0-1.0)
                0.0 = seen this before (familiar)
                1.0 = never seen anything like this (novel)
        """
        # Get or create memory for this sensor
        if sensor_id not in self.memory:
            self.memory[sensor_id] = deque(maxlen=self.memory_size)
            self._memory_timestamps[sensor_id] = deque(maxlen=self.memory_size)

        memory = self.memory[sensor_id]
        timestamps = self._memory_timestamps[sensor_id]
        now = _time.time()

        # If no history, everything is novel
        if len(memory) == 0:
            memory.append(sensor_output)
            timestamps.append(now)
            return 1.0

        # Compare to recent past experiences, weighted by recency
        samples_to_check = min(self.comparison_samples, len(memory))
        recent_mem = list(memory)[-samples_to_check:]
        recent_ts = list(timestamps)[-samples_to_check:]

        max_effective_similarity = 0.0
        for past_obs, past_ts in zip(recent_mem, recent_ts):
            raw_sim = self._compute_similarity(sensor_output, past_obs)
            # Old observations feel less familiar — their similarity pull fades
            recency = time_decay_weight(past_ts, now, self.half_life)
            effective_sim = raw_sim * recency
            if effective_sim > max_effective_similarity:
                max_effective_similarity = effective_sim

        novelty = 1.0 - max_effective_similarity

        # Store current observation
        memory.append(sensor_output)
        timestamps.append(now)

        # Clip to valid range (handle floating point precision issues)
        return float(np.clip(novelty, 0.0, 1.0))

    def _compute_similarity(self, current: Any, past: Any) -> float:
        """
        Compute similarity between observations

        Returns value 0.0-1.0:
            0.0 = completely different
            1.0 = identical
        """
        if isinstance(current, torch.Tensor) and isinstance(past, torch.Tensor):
            # Cosine similarity for tensors
            current_flat = current.flatten()
            past_flat = past.flatten()

            if current_flat.shape != past_flat.shape:
                # Different shapes - not similar
                return 0.0

            cos_sim = torch.nn.functional.cosine_similarity(
                current_flat.unsqueeze(0),
                past_flat.unsqueeze(0)
            ).item()

            # Convert from [-1, 1] to [0, 1]
            return (cos_sim + 1.0) / 2.0

        elif isinstance(current, np.ndarray) and isinstance(past, np.ndarray):
            # Cosine similarity for numpy arrays
            current_flat = current.flatten()
            past_flat = past.flatten()

            if current_flat.shape != past_flat.shape:
                return 0.0

            dot = np.dot(current_flat, past_flat)
            norm_product = np.linalg.norm(current_flat) * np.linalg.norm(past_flat)

            if norm_product == 0:
                return 0.0

            cos_sim = dot / norm_product
            return (cos_sim + 1.0) / 2.0

        elif isinstance(current, (int, float)) and isinstance(past, (int, float)):
            # Inverse normalized distance for scalars
            distance = abs(current - past)
            # Map distance to similarity (closer = more similar)
            # Using exponential decay: similarity = exp(-distance)
            similarity = np.exp(-distance / 10.0)  # Normalize by expected range
            return float(similarity)

        else:
            # Unknown types - can't compare
            return 0.5  # Neutral similarity

    def get_memory_size(self, sensor_id: str) -> int:
        """Get number of stored observations for sensor"""
        if sensor_id not in self.memory:
            return 0
        return len(self.memory[sensor_id])

    def reset_sensor(self, sensor_id: str):
        """Clear memory for specific sensor"""
        if sensor_id in self.memory:
            del self.memory[sensor_id]
        if sensor_id in self._memory_timestamps:
            del self._memory_timestamps[sensor_id]

    def reset_all(self):
        """Clear all memory"""
        self.memory.clear()
        self._memory_timestamps.clear()

    def get_most_novel_observation(self, sensor_id: str, n: int = 5) -> List[Any]:
        """
        Get the n most novel observations from history

        Useful for identifying interesting past experiences
        """
        if sensor_id not in self.memory:
            return []

        memory = list(self.memory[sensor_id])
        if len(memory) < 2:
            return memory

        # Compute novelty of each observation relative to others
        novelty_scores = []
        for i, observation in enumerate(memory):
            # Compare to all OTHER observations
            similarities = []
            for j, other in enumerate(memory):
                if i != j:
                    sim = self._compute_similarity(observation, other)
                    similarities.append(sim)

            # Novelty = 1 - max similarity to others
            max_sim = max(similarities) if similarities else 0.0
            novelty = 1.0 - max_sim
            novelty_scores.append((novelty, observation))

        # Sort by novelty (descending)
        novelty_scores.sort(reverse=True, key=lambda x: x[0])

        # Return top n
        return [obs for _, obs in novelty_scores[:n]]

File: sage-rs/sage-lib/src/snarc/novelty.rs
use std::collections::HashMap;

use crate::snarc::temporal::{default_half_life, time_decay_weight, now_secs};

/// Novelty detector — measures dissimilarity from past observations.
/// Maintains per-sensor episodic memory with time-decay weighting on similarity.
pub struct NoveltyDetector {
    memory_size: usize,
    comparison_samples: usize,
    half_life: f64,
    memory: HashMap<String, Vec<f64>>,
    timestamps: HashMap<String, Vec<f64>>,
}

impl NoveltyDetector {
    pub fn new(memory_size: usize, comparison_samples: usize, half_life: Option<f64>) -> Self {
        Self {
            memory_size,
            comparison_samples,
            half_life: half_life.unwrap_or_else(|| default_half_life("novelty")),
            memory: HashMap::new(),
            timestamps: HashMap::new(),
        }
    }

    pub fn with_defaults() -> Self {
        Self::new(1000, 50, None)
    }

    /// Compute novelty for a scalar observation. Returns 0.0 (familiar) to 1.0 (novel).
    pub fn compute(&mut self, observation: f64, sensor_id: &str) -> f64 {
        let now = now_secs();

        if !self.memory.contains_key(sensor_id) {
            self.memory.insert(sensor_id.to_string(), Vec::new());
            self.timestamps.insert(sensor_id.to_string(), Vec::new());
        }

        let mem = self.memory.get(sensor_id).unwrap();

        if mem.is_empty() {
            self.memory.get_mut(sensor_id).unwrap().push(observation);
            self.timestamps.get_mut(sensor_id).unwrap().push(now);
            return 1.0;
        }

        let n = mem.len();
        let start = if n > self.comparison_samples { n - self.comparison_samples } else { 0 };
        let ts = self.timestamps.get(sensor_id).unwrap();

        let mut max_effective_sim: f64 = 0.0;
        for i in start..n {
            let raw_sim = scalar_similarity(observation, mem[i]);
            let recency = time_decay_weight(ts[i], now, self.half_life);
            let effective = raw_sim * recency;
            if effective > max_effective_sim {
                max_effective_sim = effective;
            }
        }

        let novelty = (1.0 - max_effective_sim).clamp(0.0, 1.0);

        // Store observation
        let mem = self.memory.get_mut(sensor_id).unwrap();
        let ts = self.timestamps.get_mut(sensor_id).unwrap();
        mem.push(observation);
        ts.push(now);
        if mem.len() > self.memory_size {
            mem.remove(0);
            ts.remove(0);
        }

        novelty
    }

    pub fn memory_size_for(&self, sensor_id: &str) -> usize {
        self.memory.get(sensor_id).map_or(0, |m| m.len())
    }

    pub fn reset_sensor(&mut self, sensor_id: &str) {
        self.memory.remove(sensor_id);
        self.timestamps.remove(sensor_id);
    }

    pub fn reset_all(&mut self) {
        self.memory.clear();
        self.timestamps.clear();
    }
}

/// Scalar similarity: exp(-|a - b| / 10.0), matching Python's NoveltyDetector.
fn scalar_similarity(a: f64, b: f64) -> f64 {
    (-(a - b).abs() / 10.0).exp()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn first_observation_is_novel() {
        let mut det = NoveltyDetector::with_defaults();
        assert_eq!(det.compute(5.0, "test"), 1.0);
    }

    #[test]
    fn identical_repeat_is_familiar() {
        let mut det = NoveltyDetector::with_defaults();
        det.compute(5.0, "test");
        let n = det.compute(5.0, "test");
        // similarity = exp(0) = 1.0, novelty = 0.0
        assert!(n < 0.01, "identical repeat novelty={n}, expected ~0.0");
    }

    #[test]
    fn distant_value_is_novel() {
        let mut det = NoveltyDetector::with_defaults();
        det.compute(0.0, "test");
        let n = det.compute(100.0, "test");
        // similarity = exp(-100/10) ≈ 0.0000454, novelty ≈ 1.0
        assert!(n > 0.99, "distant novelty={n}, expected ~1.0");
    }

    #[test]
    fn sensor_independence() {
        let mut det = NoveltyDetector::with_defaults();
        det.compute(5.0, "a");
        det.compute(100.0, "b");
        // Second obs on "a" same value → familiar
        let na = det.compute(5.0, "a");
        // Second obs on "b" same value → familiar
        let nb = det.compute(100.0, "b");
        assert!(na < 0.01);
        assert!(nb < 0.01);
    }

    #[test]
    fn memory_eviction() {
        let mut det = NoveltyDetector::new(5, 5, Some(300.0));
        for i in 0..10 {
            det.compute(i as f64, "test");
        }
        assert_eq!(det.memory_size_for("test"), 5);
    }

    #[test]
    fn similarity_function() {
        assert!((scalar_similarity(5.0, 5.0) - 1.0).abs() < 1e-12);
        assert!((scalar_similarity(0.0, 10.0) - (-1.0_f64).exp()).abs() < 1e-12);
    }
}
